File: backend/app/schemas/expense.py
"""Request and response contracts for expenses, balances and settlements."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class ShareInput(BaseModel):
    user_id: UUID
    share_cents: int = Field(ge=0)


class ExpenseCreate(BaseModel):
    """Two ways to split, one field decides.

    Omitting `shares` splits evenly across `participants`; providing it allows
    any uneven split. Supporting both from day one avoids a schema change once
    someone inevitably wants to pay a bigger slice.
    """

    description: str = Field(min_length=1, max_length=200)
    amount_cents: int = Field(gt=0)
    paid_by: UUID
    spent_at: datetime | None = None
    participants: list[UUID] | None = None
    shares: list[ShareInput] | None = None

    @model_validator(mode="after")
    def check_split(self):
        if self.shares is None and not self.participants:
            raise ValueError("provide either participants or shares")
        if self.shares is not None:
            if sum(s.share_cents for s in self.shares) != self.amount_cents:
                raise ValueError("shares must sum to amount_cents")
            ids = [s.user_id for s in self.shares]
            if len(ids) != len(set(ids)):
                raise ValueError("duplicate user in shares")
        if self.participants is not None and len(self.participants) != len(set(self.participants)):
            raise ValueError("duplicate participant")
        return self


class SharePublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: UUID
    share_cents: int


class ExpensePublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    trip_id: UUID
    description: str
    amount_cents: int
    currency: str
    paid_by: UUID
    spent_at: datetime
    created_by: UUID
    created_at: datetime
    shares: list[SharePublic]


class BalanceEntry(BaseModel):
    user_id: UUID
    # Positive: the group owes them. Negative: they owe the group.
    balance_cents: int


class TransferSuggestion(BaseModel):
    from_user_id: UUID
    to_user_id: UUID
    amount_cents: int


class BalanceReport(BaseModel):
    balances: list[BalanceEntry]
    suggested_transfers: list[TransferSuggestion]
    total_spent_cents: int


class SettlementCreate(BaseModel):
    to_user_id: UUID
    amount_cents: int = Field(gt=0)
    note: str | None = Field(default=None, max_length=200)


class SettlementPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    trip_id: UUID
    from_user_id: UUID
    to_user_id: UUID
    amount_cents: int
    note: str | None
    settled_at: datetime

File: mobile/lib/features/trips/data/expense.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'expense.freezed.dart';
part 'expense.g.dart';

@freezed
abstract class ExpenseShare with _$ExpenseShare {
  const factory ExpenseShare({
    required String userId,
    required int shareCents,
  }) = _ExpenseShare;

  factory ExpenseShare.fromJson(Map<String, dynamic> json) =>
      _$ExpenseShareFromJson(json);
}

@freezed
abstract class Expense with _$Expense {
  const factory Expense({
    required String id,
    required String tripId,
    required String description,
    required int amountCents,
    required String currency,
    required String paidBy,
    required DateTime spentAt,
    required String createdBy,
    required DateTime createdAt,
    required List<ExpenseShare> shares,
  }) = _Expense;

  const Expense._();

  /// What this expense costs [userId], or null when they took no part in it.
  ///
  /// Distinguishing "your share is zero" from "you weren't involved" matters:
  /// the list shows the two differently.
  int? shareFor(String userId) {
    for (final share in shares) {
      if (share.userId == userId) return share.shareCents;
    }
    return null;
  }

  factory Expense.fromJson(Map<String, dynamic> json) =>
      _$ExpenseFromJson(json);
}

@freezed
abstract class BalanceEntry with _$BalanceEntry {
  const factory BalanceEntry({
    required String userId,
    required int balanceCents,
  }) = _BalanceEntry;

  factory BalanceEntry.fromJson(Map<String, dynamic> json) =>
      _$BalanceEntryFromJson(json);
}

@freezed
abstract class TransferSuggestion with _$TransferSuggestion {
  const factory TransferSuggestion({
    required String fromUserId,
    required String toUserId,
    required int amountCents,
  }) = _TransferSuggestion;

  factory TransferSuggestion.fromJson(Map<String, dynamic> json) =>
      _$TransferSuggestionFromJson(json);
}

@freezed
abstract class BalanceReport with _$BalanceReport {
  const factory BalanceReport({
    required List<BalanceEntry> balances,
    required List<TransferSuggestion> suggestedTransfers,
    required int totalSpentCents,
  }) = _BalanceReport;

  const BalanceReport._();

  int balanceFor(String userId) {
    for (final entry in balances) {
      if (entry.userId == userId) return entry.balanceCents;
    }
    return 0;
  }

  factory BalanceReport.fromJson(Map<String, dynamic> json) =>
      _$BalanceReportFromJson(json);
}
